### PR TITLE
Dev/270 [PUT] /api/problems/{problemId}/{langEnv}/providedCodes

### PR DESCRIPTION
**Pull request by Chaoyuuu**
# Description

See the linked issues for the details.

- If you have other implementation details, write it down here.

### CODE REVIEW

File: Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/repositories/ProblemRepository.java [hunk 1]
```diff
@@ -49,6 +50,8 @@
 
     void patchProblem(int problemId, PatchProblemParams params);
 
+    String updateProblemWithProvidedCodes(Problem problem, Language language, List<FileResource> providedCodes);
```
Comment (Johnny850807) [inline Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/repositories/ProblemRepository.java:53]: `void updateProblemWithProvidedCodes(Problem problem, Language language, List<FileResource> providedCodes);`

File: Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/usecases/BaseProblemUseCase.java [hunk 2]
```diff
@@ -29,6 +29,6 @@ public BaseProblemUseCase(ProblemRepository problemRepository) {
 
     protected Problem doFindProblemById(int problemId) throws NotFoundException {
         return problemRepository.findProblemById(problemId)
-                .orElseThrow(() -> new NotFoundException(problemId, "problem"));
+                .orElseThrow(() -> NotFoundException.notFound(Problem.class).id(problemId));
```
Comment (Johnny850807) [inline Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/usecases/BaseProblemUseCase.java:32]: static import

File: Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/usecases/UploadProvidedCodeUseCase.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    public static final String PROVIDED_CODE_MULTIPART_KEY_NAME = "providedCodes";
+
+    public UploadProvidedCodeUseCase(ProblemRepository problemRepository) {
+        super(problemRepository);
+    }
+
+    public void execute(Request request, Presenter presenter) {
+        Problem problem = doFindProblemById(request.problemId);
+        String fileId = problemRepository.updateProblemWithProvidedCodes(problem, request.language, request.providedCodes);
+        presenter.setProvidedCodesFileId(fileId);
```
Comment (Johnny850807) [inline Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/usecases/UploadProvidedCodeUseCase.java:39]: `presenter.showResult(problem);`

File: Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/usecases/UploadProvidedCodeUseCase.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+@Named
+public class UploadProvidedCodeUseCase extends BaseProblemUseCase {
+    public static final String PROVIDED_CODE_MULTIPART_KEY_NAME = "providedCodes";
+
+    public UploadProvidedCodeUseCase(ProblemRepository problemRepository) {
+        super(problemRepository);
+    }
+
+    public void execute(Request request, Presenter presenter) {
+        Problem problem = doFindProblemById(request.problemId);
```
Comment (Wally5077) [inline Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/usecases/UploadProvidedCodeUseCase.java:37]: Problem problem = findProblem(request.problemId);

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java [hunk 5]
```diff
@@ -197,6 +199,47 @@ public void saveTags(List<String> tagList) {
...
+        return fileId;
+    }
+
+    private String saveProvidedCodesAndGetFileId(int problemId, Language language, List<FileResource> providedCodes) {
+        String fileName = format("%d-%s-provided.zip", problemId, language.toString());
+        ByteArrayInputStream zip = zipToStream(providedCodes);
+        return gridFsTemplate.store(zip, fileName).toString();
+    }
+
+    private void updateProvidedCodesFileIdInProblem(Problem problem, String fileId, Language language) {
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java:215]: order consistency : `updateProvidedCodesFileIdInProblem(Problem problem, Language language, String fileId)`

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 6]
```diff
@@ -528,5 +538,74 @@ private Problem saveProblemAndGet(int problemId) {
         saveProblems(problemId);
         return problemRepository.findProblemById(problemId).orElseThrow();
     }
+
+    @Test
+    void GivenTwoProvidedCodes_WhenUploadProvidedCodes_ShouldResponseProvidedCodesFileId() throws Exception {
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:543]: `GiveOneProblemSaved_WhenUploadTwoProvidedCodes_ShouldRespondProvidedCodesFileId`

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 7]
```diff
@@ -528,5 +538,74 @@ private Problem saveProblemAndGet(int problemId) {
...
+        Language language = Language.C;
+        int problemId = 1;
+        saveProblems(problemId);
+
+        String fileId = uploadProvidedCodesAndGetFileId(problemId, language, getTwoProvidedCodes());
+
+        problemRepository.findProblemById(problemId)
+                .ifPresent(problem -> assertFileIdEquals(fileId, problem, language));
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:551]:
```java
Problem problem = getProblem(problemId);
problemShouldHaveProvidedCodesId(problem, fileId);
```

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 8]
```diff
@@ -528,5 +538,74 @@ private Problem saveProblemAndGet(int problemId) {
...
+    }
+
+    @Test
+    void GivenNotExistedProblemIdAndTwoProvidedCodes_WhenUploadProvidedCodes_ShouldResponseNotFound() throws Exception {
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:555]: improve this

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 9]
```diff
@@ -528,5 +538,74 @@ private Problem saveProblemAndGet(int problemId) {
...
+    void GivenProblemWithEmptyLanguageEnvAndTwoProvidedCodes_WhenUploadProvidedCodes_ShouldResponseProvidedCodesFileIds() throws Exception {
+        Language language = Language.C;
+        int problemId = saveProblemWithTitle("problemTitle");
+
+        String fileId = uploadProvidedCodesAndGetFileId(problemId, language, getTwoProvidedCodes());
+
+        problemRepository.findProblemById(problemId)
+                .ifPresent(problem -> assertFileIdEquals(fileId, problem, language));
+    }
+
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:572]: ditto

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

**Review by Wally5077 — APPROVED** [review]
LGTM

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 10]
```diff
@@ -528,5 +538,77 @@ private Problem saveProblemAndGet(int problemId) {
...
+    @Test
+    void WhenUploadTwoProvidedCodesWithNonExistingProblemId_ShouldResponseNotFound() throws Exception {
+        Language language = Language.C;
+        int nonExistingProblemId = 123;
+
+        uploadProvidedCodes(nonExistingProblemId, language, getTwoProvidedCodes()).andExpect(status().isNotFound());
+    }
+
+    @Test
+    void GiveOneProblemSavedWithNonExistingLanguageEnv_WhenUploadTwoProvidedCodes_ShouldResponseProvidedCodesFileIds() throws Exception {
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:563]: improve, this is not the given

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 11]
```diff
@@ -528,5 +538,77 @@ private Problem saveProblemAndGet(int problemId) {
...
+        saveProblems(problemId);
+
+        String fileId = uploadProvidedCodesAndGetFileId(problemId, language, getTwoProvidedCodes());
+
+        ProblemView problem = getProblem(problemId);
+        problemShouldHaveProvidedCodesId(problem, fileId, language);
+    }
+
+    @Test
+    void WhenUploadTwoProvidedCodesWithNonExistingProblemId_ShouldResponseNotFound() throws Exception {
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:555]: response?

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 12]
```diff
@@ -528,5 +538,77 @@ private Problem saveProblemAndGet(int problemId) {
...
+        Language language = Language.C;
+        int nonExistingProblemId = 123;
+
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:558]: no spaces

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 13]
```diff
@@ -528,5 +538,77 @@ private Problem saveProblemAndGet(int problemId) {
...
+                .andExpect(status().isOk()));
+    }
+
+    private ResultActions uploadProvidedCodes(int problemId, Language language, MockMultipartFile[] files) throws Exception {
+        return mockMvc.perform(multipartRequestWithProvidedCodes(problemId, language, files));
+    }
+
+    private MockHttpServletRequestBuilder multipartRequestWithProvidedCodes(int problemId, Language language, MockMultipartFile... files) {
+        var call =
+                multipart("/api/problems/{problemId}/{langEnvName}/providedCodes", problemId, language.toString());
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:593]: format

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java [hunk 14]
```diff
@@ -197,6 +199,47 @@ public void saveTags(List<String> tagList) {
...
+                            .language(language)
+                            .providedCodesFileId(fileId)
+                            .build();
+                    problem.addLanguageEnv(langEnv);
+                    updateLanguageEnv(problem.getId(), langEnv, language);
+                });
+    }
+
+    private void removeFileIdIfExist(String fileId) {
+        if (!fileId.isEmpty()) {
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java:231]: use `StringUtils.isNullOrEmpty` with static import

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java [hunk 15]
```diff
@@ -197,6 +199,47 @@ public void saveTags(List<String> tagList) {
...
+                    LanguageEnv langEnv = LanguageEnv.builder()
+                            .language(language)
+                            .providedCodesFileId(fileId)
+                            .build();
+                    problem.addLanguageEnv(langEnv);
+                    updateLanguageEnv(problem.getId(), langEnv, language);
+                });
+    }
+
+    private void removeFileIdIfExist(String fileId) {
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java:230]: rename `removeFileIdIfExists`

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java [hunk 16]
```diff
@@ -197,6 +199,47 @@ public void saveTags(List<String> tagList) {
...
+                    updateLanguageEnv(problem.getId(), langEnv, language);
+                }, () -> {
+                    LanguageEnv langEnv = LanguageEnv.builder()
+                            .language(language)
+                            .providedCodesFileId(fileId)
+                            .build();
+                    problem.addLanguageEnv(langEnv);
+                    updateLanguageEnv(problem.getId(), langEnv, language);
+                });
+    }
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java:228]: create an issue and add TODO comment for this:
Transaction: Operations in `updateProvidedCodesFileIdInProblem` should be atomic

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 17]
```diff
@@ -552,15 +552,14 @@ void GiveOneProblemSaved_WhenUploadTwoProvidedCodes_ShouldRespondProvidedCodesFi
...
+    void WhenUploadTwoProvidedCodesWithNonExistingProblemId_ShouldRespondNotFound() throws Exception {
         Language language = Language.C;
         int nonExistingProblemId = 123;
-
         uploadProvidedCodes(nonExistingProblemId, language, getTwoProvidedCodes()).andExpect(status().isNotFound());
     }
 
     @Test
-    void GiveOneProblemSavedWithNonExistingLanguageEnv_WhenUploadTwoProvidedCodes_ShouldResponseProvidedCodesFileIds() throws Exception {
+    void GiveOneProblemSavedWithoutLanguageEnv_WhenUploadTwoProvidedCodes_ShouldRespondProvidedCodesFileIds() throws Exception {
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:562]: `GiveOneProblemSavedWithoutLanguageEnv_WhenUploadTwoProvidedCodes_ShouldRespondProvidedCodesFileId`

**Review by Johnny850807 — APPROVED** [review]
Last but not least

**Pull request merged**